File: 53-maximum-subarray/53-maximum-subarray.py
# DP
# class Solution:
#     def maxSubArray(self, nums: List[int]) -> int:
#         dp = [0]*len(nums)
#         dp[0] = nums[0]
#         for i in range(1, len(nums)):
#             dp[i] = max(dp[i-1] + nums[i], nums[i])
#         return max(dp)

# KADANE'S ALGO
   
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        currSum = maxSum = nums[0]
        for i in range(1,len(nums)):
            currSum += nums[i]
            currSum = max(currSum, nums[i])
            maxSum  = max(maxSum, currSum)
            
        return maxSum

File: 53-maximum-subarray/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// Kadane's algorithm
const maxSubArray = (nums) => {
    if (!nums || nums.length === 0) return 0;
    let cur = nums[0];
    let max = nums[0];
    for (let i = 1; i < nums.length; i++) {
        cur = Math.max(nums[i], cur + nums[i]);
        max = Math.max(max, cur);
    }
    return max;
};